### DISCUSSION TITLE
Operator overloading

### DISCUSSION THREAD

**Discussion by data-man** [c3lang/c3c#432] [Polls]
Polls testing

**Comment by ejv2** [comment]
I think that operator overloading severely harms the clarity of most languages that have it. Given that C3 already has functions that can be called on structs using the dot (".") operator, I think that this would be a misfeature and a negative change.

**Comment by mtrudeau** [comment]
Specialized operator overloading is sometimes very convenient and makes code far easier to read.

Thinking mostly about [] overloading for vectors and maps.   

m["Hello"] = 5    looks nicer to write and read than    m.put("Hello", 5);

**Reply by lerno** [reply]
This already works.

**Comment by helikopterodaktyl** [comment]
Operator overloading is a big improvement for math libraries, especially in the gamedev fields. Similarly types such as complex numbers also benefit.

v1 + (v2 - v3 * 0.5) + v4 * v5 looks better than v1.add(v2.sub(v3.mul(0.5)).add(v4.mul(v5))

Personally I don't think operator overloading goes against the C ideology. Usually C++ iostreams and boost::spirit are mentioned when it comes to operator overload abuse in other languages, but that's pretty much it. It's unusual in general to see operator overloading outside of math libraries and containers.

**Reply by lerno** [reply]
For C3, SIMD vector types allow you to use arithmetics directly. The operator overloading on arithmetics is less straightforward than one would think. First of all there are many different overloads that needs to be implemented, as `a + b` might not be the same as `b + a`. And then there are the combined assignment operators `a += b`. And then there is the interaction between those and subscript, e.g `foo[i] += b`. It's not a question of "is it nice to have" but "does it add enough value to the language, and what feature would you want to sacrifice to have it?"

**Comment by lerno** [comment]
Operator overloading is coming in 0.7.1 as a test 😄
